Compilation

Multimodcrafter opened the pull request.

Multimodcrafter committed:
Add new instructions to NFA

We require two vm instructions 'CheckLookaround' and 'WriteLookaround'
to be able to track the state of lookaround expressions at the
current position in the haystack. Both instructions access
a new 'lookaround' vector of booleans, which contains one entry
per lookaround expression in the regex.

Multimodcrafter committed:
Implement lookaround compilation

These changes implement the compilation of lookaround assertions from
HIR to NFA. Subexpressions of lookaround assertions are patched to
a top level reverse union. This is necessary so that the NFA will
explore the innermost subexpression first and thereby make sure that
all subexpression results are available when they need to be checked.
I.e. any `WriteLookaround` state must be visited before any
`CheckLookaround` state with the same index.

shilangyu commented on regex-automata/src/nfa/thompson/backtrack.rs:
```diff
@@ -1519,6 +1519,10 @@ impl BoundedBacktracker {
                     }
                     sid = next;
                 }
+                State::WriteLookaround { .. }
+                | State::CheckLookaround { .. } => {
+                    todo!("check how to handle")
+                }
```
Hmm this might be quite unfortunate. The backtracking engine should be able to handle lookbehinds easily, but will it be able to do so with these vm commands?

Multimodcrafter replied:
That is indeed something we need to dig into. Let's discuss it in the upcoming meeting.

shilangyu commented on regex-automata/src/nfa/thompson/builder.rs:
```diff
@@ -91,10 +98,20 @@ enum State {
         /// The next state that this state should transition to.
         next: StateID,
     },
+    WriteLookaround {
+        lookaround_index: usize,
+    },
+    CheckLookaround {
+        lookaround_index: usize,
+        positive: bool,
+        next: StateID,
+    },
```
Shouldn't we name those `WriteLookBehind` and `CheckLookBehind`? Since this does not extend to look-aheads

Multimodcrafter replied:
IIRC, the lookahead implementation uses the same instructions, but with a table that has an entry for each string position, so it should be possible to use these instructions for lookaheads as well.

shilangyu commented on regex-automata/src/nfa/thompson/builder.rs:
```diff
@@ -91,10 +98,20 @@ enum State {
         /// The next state that this state should transition to.
         next: StateID,
     },
+    WriteLookaround {
+        lookaround_index: usize,
+    },
+    CheckLookaround {
+        lookaround_index: usize,
+        positive: bool,
+        next: StateID,
+    },
```
There is missing documentation

Multimodcrafter replied:
True, for public items, the compiler denies missing docs, but these are private. We should definitely document these anyway :)

shilangyu commented on regex-automata/src/nfa/thompson/builder.rs:
```diff
@@ -693,6 +732,30 @@ impl Builder {
         self.add(State::Empty { next: StateID::ZERO })
     }
 
+    /// Add a state which will record that the lookaround with the given index
+    /// is satisfied at the current position.
+    pub fn add_write_lookaround(
```
Here too I would probably refer to look-behinds only

shilangyu commented on regex-automata/src/nfa/thompson/builder.rs:
```diff
@@ -91,10 +98,20 @@ enum State {
         next: StateID,
     },
+    WriteLookaround {
+        lookaround_index: usize,
```
We probably want to use `SmallIndex` here

Multimodcrafter replied:
Yeah I thought about that as well

shilangyu commented on regex-automata/src/nfa/thompson/nfa.rs:
```diff
@@ -1260,6 +1266,7 @@ pub(super) struct Inner {
     /// zero-length prefix for any of the patterns in this NFA.
     look_set_prefix_all: LookSet,
     */
+    look_count: usize,
```
I would name it `lookaround_count` to avoid confusion with the assertion looks

Multimodcrafter replied:
good point

shilangyu reviewed (commented):
Many tests seem to be failing, do we want to adjust them?

I did not finish reading the PR yet, but I thought I would leave some intermediate comments

Multimodcrafter commented:
> Many tests seem to be failing, do we want to adjust them?

which tests are you talking about? If I run `./test` locally, everything passes...

Multimodcrafter committed:
Restore compilation behaviour for regexes without lookarounds

The machinery necessary to perform the parallel lookbehind checking
should only be compiled in when there is actually a lookbehind expression
in the regex. This restores compilation to the expected outputs for
regexes without lookbehind expressions.

Multimodcrafter committed:
Address review comments

Multimodcrafter committed:
Implement look-around index generation

Multimodcrafter committed:
Change tracking of look-around state to index

This makes it so we don't need to reset the lookaround state on
each character advancement.

shilangyu commented on regex-cli/cmd/generate/fowler.rs:
```diff
@@ -418,5 +418,6 @@ fn count_capturing_groups_ast(ast: &regex_syntax::ast::Ast) -> usize {
         Ast::Concat(ref concat) => {
             concat.asts.iter().map(count_capturing_groups_ast).sum()
         }
+        Ast::LookAround(_) => todo!(),
```
Sorry for that, fixed in #4

shilangyu commented on regex-automata/src/nfa/thompson/pikevm.rs:
```diff
@@ -1902,6 +1988,11 @@ impl Cache {
             stack: vec![],
             curr: ActiveStates::new(re),
             next: ActiveStates::new(re),
+            lookaround: {
+                let mut res = Vec::new();
+                res.resize(re.lookaround_count().as_usize(), None);
+                res
+            },
```
Any particular reason for not using this?
```suggestion
            lookaround: vec![None; re.lookaround_count().as_usize()],
```

Multimodcrafter replied:
Yes, it is called "ignorance", hahaha 😝

Multimodcrafter committed:
Fix cli tool and AST->HIR translation

Multimodcrafter committed:
Fix lookaround union order

shilangyu commented on regex-automata/src/nfa/thompson/builder.rs:
```diff
@@ -91,6 +91,17 @@ enum State {
     },
+    /// An empty state that behaves analogously to a `Match` state but for
+    /// the look-around sub-expression with the given index.
+    WriteLookaround { lookaround_index: SmallIndex },
```
For consistency with the rest of the codebase we should probably name it `WriteLookAround` (same goes to `CheckLookAround`)

shilangyu commented on regex-automata/src/nfa/thompson/builder.rs:
```diff
@@ -91,6 +91,17 @@ enum State {
+    /// the look-around sub-expression with the given index.
+    WriteLookaround { lookaround_index: SmallIndex },
+    /// A conditional epsilon transition that will only be taken if the
+    /// look-around sub-expression with the given index evaluates to `positive`
+    /// at the current position in the haystack.
```
That is outdated now that we use indices

Multimodcrafter replied:
Not really, we are using indices to track the state of look-arounds but the semantics of the `CheckLookAround` instruction is still to provide a conditional epsilon transition on the predicate as described. That is, the transition will be taken if the look-around sub-expression evaluates to `positive` at the current position, but the way this is checked is no longer via direct boolean comparison but rather by first evaluating `last_true_pos == current_pos` and only then compare the result to `positive`.

shilangyu replied:
I was referring to the point about how this check is now done. But indeed the docs do not have to be this precise about it here

shilangyu commented on regex-automata/src/nfa/thompson/compiler.rs:
```diff
@@ -1003,14 +1033,43 @@ impl Compiler {
+        &self,
+        lookaround: &LookAround,
+    ) -> Result<ThompsonRef, BuildError> {
+        let sub = match lookaround {
+            LookAround::NegativeLookBehind(ref sub)
+            | LookAround::PositiveLookBehind(ref sub) => self.c(sub)?,
+        };
```
```suggestion
        let sub = self.c(lookaround.sub())?;
```

shilangyu commented on regex-automata/src/nfa/thompson/compiler.rs:
```diff
@@ -1003,14 +1033,43 @@ impl Compiler {
+        self.patch(
+            self.lookaround_alt
+                .borrow()
+                .expect("Cannot compile lookaround outside pattern"),
```
```suggestion
                .expect("Cannot compile look-around outside pattern"),
```

shilangyu commented on regex-automata/src/nfa/thompson/nfa.rs:
```diff
@@ -1100,6 +1100,12 @@ impl NFA {
         self.0.look_set_prefix_any
     }
 
+    /// Returns how many lookaround sub-expressions this nfa contains
```
```suggestion
    /// Returns how many look-around sub-expressions this nfa contains.
```

shilangyu commented on regex-automata/src/nfa/thompson/nfa.rs:
```diff
@@ -1545,6 +1563,25 @@ pub enum State {
+    /// This is like a match state but for a lookaround expression
+    /// executing this state will write a `true` into the lookaround oracle at
+    /// index `look_idx`
+    WriteLookaround {
```
Same with renaming

shilangyu commented on regex-automata/src/nfa/thompson/nfa.rs:
```diff
@@ -1545,6 +1563,25 @@ pub enum State {
         /// satisfied.
         next: StateID,
     },
+    /// This is like a match state but for a lookaround expression
+    /// executing this state will write a `true` into the lookaround oracle at
+    /// index `look_idx`
+    WriteLookaround {
+        /// The index of the lookaround expression that matches
+        look_idx: SmallIndex,
+    },
```
Sorry for being so nit-picky about naming. The codebase seems to follow these conventions. 

```suggestion
    /// This is like a match state but for a look-around expression
    /// executing this state will write a `true` into the look-around oracle at
    /// index `look_idx`.
    WriteLookaround {
        /// The index of the look-around expression that matches.
        lookaround_idx: SmallIndex,
    },
```

Multimodcrafter replied:
here, in fact, the documentation is no longer correct :)

shilangyu commented on regex-automata/src/nfa/thompson/nfa.rs:
```diff
@@ -1545,6 +1563,25 @@ pub enum State {
+    /// If `positive` is false, then the lookaround expression is negative and
+    /// hence must NOT hold.
+    CheckLookaround {
+        /// The index of the lookaround expression that must be satisfied
+        look_idx: SmallIndex,
+        /// Whether this is a positive lookaround expression
+        positive: bool,
+        /// The next state to transition if the lookaround assertion is satisfied
```
```suggestion
        /// The index of the look-around expression that must be satisfied.
        lookaround_idx: SmallIndex,
        /// Whether this is a positive look-around expression.
        positive: bool,
        /// The next state to transition if the lookaround assertion is satisfied.
```

shilangyu commented on regex-automata/src/nfa/thompson/nfa.rs:
```diff
@@ -1748,6 +1793,18 @@ impl fmt::Debug for State {
             State::Look { ref look, next } => {
                 write!(f, "{:?} => {:?}", look, next.as_usize())
             }
+            State::WriteLookaround { look_idx } => {
+                write!(f, "Write Lookaround: {}", look_idx.as_u32())
+            }
+            State::CheckLookaround { look_idx, positive, next } => {
+                write!(
+                    f,
+                    "Check Lookaround {} is {} => {}",
+                    look_idx.as_u32(),
+                    positive,
+                    next.as_usize()
+                )
```
We probably want to adjust to their style?

```suggestion
            State::WriteLookaround { look_idx } => {
                write!(f, "write-look-around({})", look_idx.as_u32())
            }
            State::CheckLookaround { look_idx, positive, next } => {
                write!(
                    f,
                    "check-look-around({} is {}) => {}",
                    look_idx.as_u32(),
                    if positive { "matched" } else { "not matched" },
                    next.as_usize()
                )
```

shilangyu commented on regex-syntax/src/hir/mod.rs:
```diff
@@ -2134,6 +2135,17 @@ impl Properties {
+    ///
+    /// Only returns true for [`HirKind::LookAround`] and not for
+    /// [`HirKind::Look`], which can be queried by [`look_set`] instead.
+    /// Currently, only lookbehind assertions without capture groups are
+    /// supported.
```
I would not mention this here. This is not the concern of `Properties` and will make it harder to update the docs when the support is changed.

Multimodcrafter committed:
Address review comments

Rename certain enums to be consistent with rest of codebase.
